compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp

```diff
@@ -172,45 +290,271 @@ static void padConvOp(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
-    Operation *rootOp = getOperation();
-    ArrayAttr mmaKinds = nullptr;
-    for (IREE::HAL::ExecutableTargetAttr targetAttr :
-         IREE::HAL::DeviceTargetAttr::lookupExecutableTargets(rootOp)) {
-      FailureOr<ArrayAttr> candidateMmaKinds =
-          getSupportedMmaTypes(targetAttr.getConfiguration());
-      if (succeeded(candidateMmaKinds)) {
-        mmaKinds = *candidateMmaKinds;
-        break;
+static void padContractionLikeOp(RewriterBase &rewriter,
+                                 linalg::LinalgOp linalgOp) {
+  FailureOr<mlir::linalg::ContractionDimensions> contractionDims =
+      mlir::linalg::inferContractionDims(linalgOp);
+
+  if (failed(contractionDims)) {
+    return;
+  }
+
+  if (contractionDims->k.size() < 1 || contractionDims->m.size() < 1 ||
+      contractionDims->n.size() < 1) {
+    return;
+  }
+
+  // Early exit if cannot find intrinsics or if multiple executable targets.
+  SmallVector<GPUMatmulShapeType> intrinsics = getIntrinsics(linalgOp);
+  if (intrinsics.empty())
+    return;
+
+  Location loc = linalgOp.getLoc();
+
+  // Naive handling by only looking into most inner dimensions.
+  int64_t mDim = contractionDims->m.back();
+  int64_t nDim = contractionDims->n.back();
+  int64_t kDim = contractionDims->k.back();
+
+  // If none of the shape is dynamic, we'd fallback to using pad to intrinsics.
+  SmallVector<int64_t, 4> bounds = linalgOp.getStaticLoopRanges();
+  int64_t mSize = bounds[mDim];
+  int64_t nSize = bounds[nDim];
+  int64_t kSize = bounds[kDim];
+
+  // Bail out on matvec-like cases.
+  if (mSize == 1 || nSize == 1) {
+    return;
+  }
+
+  // Util to get operand and dim;
+  auto getSrcOperandAndDim =
+      [&](int64_t targetDim) -> std::optional<std::pair<Value, int64_t>> {
+    for (auto operand : linalgOp.getDpsInputOperands()) {
+      auto operandMap = linalgOp.getMatchingIndexingMap(operand);
+      std::optional<unsigned> maybeDim = operandMap.getResultPosition(
+          getAffineDimExpr(targetDim, operandMap.getContext()));
+      if (maybeDim)
+        return std::make_pair(operand->get(), maybeDim.value());
+    }
+    return std::nullopt;
+  };
+
+  // Utils to compute padding size.
+  OpFoldResult zero = rewriter.getIndexAttr(0);
+  AffineExpr s0, s1; // problemSize, intrinsicSize
+  bindSymbols(rewriter.getContext(), s0, s1);
+  AffineExpr padByExpr = (s0).ceilDiv(s1) * s1 - s0;
+  auto getPadding = [&](OpFoldResult value, int64_t padTo) {
+    return affine::makeComposedFoldedAffineApply(
+        rewriter, loc, padByExpr, {value, rewriter.getIndexAttr(padTo)});
+  };
+
+  SmallVector<SmallVector<std::pair<int64_t, int64_t>>> dimsToExpandCandidates;
+  SmallVector<SmallVector<int64_t>> expandSizesCandidates;
+  SmallVector<std::array<OpFoldResult, 3>> mnkPaddingCandidates;
+  // Compute Pad and Expand metadata for linalgOp's M, N, K dimensions. Multiple
+  // candidates of the metadata will be formed based on different intrinsics.
+  for (auto &intrinsic : intrinsics) {
```

Comment:
nit: write the type when it's not obvious by looking at the RHS

Reply:
done!